homeassistant/components/upnp/__init__.py

```diff
@@ -86,17 +72,43 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up UPnP/IGD device from a config entry."""
     _LOGGER.debug("Setting up config entry: %s", entry.unique_id)
 
-    # Discover and construct.
     udn = entry.data[CONFIG_ENTRY_UDN]
     st = entry.data[CONFIG_ENTRY_ST]  # pylint: disable=invalid-name
+    usn = f"{udn}::{st}"
+
+    # Register device discovered-callback.
+    device_found_event = asyncio.Event()
+    discovery_info: Optional[Mapping[str, Any]] = None
+
+    @callback
+    def device_found(info: Mapping[str, Any]) -> None:
+        nonlocal discovery_info
+        _LOGGER.info(
+            "Device discovered: %s, at: %s", usn, info[ssdp.ATTR_SSDP_LOCATION]
+        )
+        discovery_info = info
+        device_found_event.set()
+
+    entry.async_on_unload(
```

Comment:
Won't this trigger (harmless) calls when the same device is discovered at a different location? I'd rather have a once-callback instead, if possible

Reply:
Actually, this is for the line before, referring `ssdp.async_register_callback()`.

Reply:
`ssdp.async_register_callback` returns a callable that can be used to cancel it so you could cancel it as soon as the device is found or unload.